WIP: testing on windows and conda packages/ wheels for master

JanSchulz opened the pull request:
Current Todo:
* [ ] switch addon channel to janschulz to get functools32 and msinttypes
* [ ] Update meta.yml adnd build.sh with versions from my package builder
* [ ] remove the py27 restriction in appveyor.yml after the channel switch
* [ ] write short instruction how to compile on windows with conda (basically it should be "add the janschulz conda channel, install the following libs, copy some libs, run ms scripts on some python versions, run setup..." -> what appveyor does + the run_in_env command...)
* [ ] investigate if the copy step can be replaced with runtime depencies... the conda matplotlib package suggests that this works...
* [ ] look at https://github.com/conda-forge/staged-recipes/pull/3, maybe there is something in there which I haven't found...
* [x] make appveyor fail if the tests fail...
* [x] investigate the py35 failure...
* [x] split out the wheel workaround into it's own commit
* [x] decide what to do about tkagg -> is that needed? -> would need a installer for the tcl files (see conda bld.bat)
* [x] investigate what changed in the conda freetype package >= 2.6 -> checked for file is missing there
* [x] and move travis helper scripts to ci
* [x] build a functools32 package until conda has such a package available... Until then: no py27 conda packages...
* [x] now the include dirs get included twice?
* [x] add a rule to zip up the resulting images to investigate the image test failures
* [x] Finally: got through the tests failures and open issues for them... 
* [x] get it to run...
* [x] re-enable travis 
* [x] add python dirs to default include dirs to make packaging on conda easier -> extra commit/PR
* [x] add to the build matrix: 64bit Py35 34, 27 at the first position, py27, 35 32bit last

See also the discussion at https://github.com/matplotlib/matplotlib/issues/5595

Closes: https://github.com/matplotlib/matplotlib/issues/5595
Closes: https://github.com/matplotlib/matplotlib/issues/5610 # test failures on windows
Closes: https://github.com/matplotlib/matplotlib/pull/5460 # PR which fixes a testfailure originally reported in https://github.com/matplotlib/matplotlib/issues/5446, merged into this PR 
Closes: https://github.com/matplotlib/matplotlib/pull/5278 # namedTempfile on windows problem

JanSchulz commented:
Not that I currently `amend` and `push -f` until it works (which up to now it doesn't :-()

mdboom commented on tools/appveyor_install.ps1:
```diff
@@ -0,0 +1,96 @@
+# Sample script to install Miniconda under Windows
```
Just in terms of organization, we already have some Travis-specific tools in `.travis`.

I think I'd prefer to see this sort of structure:

```
integration
+- travis (existing content in .travis)
+- appveyor
+- conda
```

and leave `tools` for just developer tools as it is now.

JanSchulz replied:
Can do, I just didn't want to add to much to the main dir...

mdboom replied:
Agreed.  I think we'd just be adding one new top-level directory (`integration`) under which all of this CI/build management stuff would live.  Probably should vendor our .deb and .rpm specs there, too someday.

JanSchulz replied:
I use "ci" which I found easier to identify than "integration" (It's also used be pandas)...

JanSchulz replied:
didn't yet push the travis stuff to that dir...

mdboom commented on patched_bdist_wheel.py:
```diff
@@ -0,0 +1,103 @@
+"""
```
This should probably go somewhere other than the top-level as well.

JanSchulz replied:
this would actually be used by all people on windows, not only CI: If I understand that bugreport right, then this patch will be needed if you want to build a wheel on windows... Not sure where this should go, tools? If it's in a subdir, it also needs a `__init__.py`?

mdboom replied:
Perhaps it needs a better name, but I had thought `integration` would be for all build-related stuff.  `tools` is really developer tools.

JanSchulz replied:
my reasoning: it's similar to setupext.py, so should be next to it. I could add that there or name it `setup_patches.py`?

JanSchulz commented:
I will remove `.travis.yml` in the next push to not swamp travis with the repushes...

tacaswell commented:
Is there a flag on circleci that someone with admin rights needs to flip?

JanSchulz commented:
On appveyor: someone with owner/admin rights on the github matplotlib org needs to login there and enable it.

mdboom commented:
@tacaswell: Do you mean circleci or appveyor?  I experimented with a bunch of these at one time, so I wouldn't be surprised if we have accounts a number of places already.

@JanSchulz: I'll try to enable appveyor.

mdboom commented:
I added matplotlib/matplotlib as a repo on appveyor.  Not sure how to fire this off again, though.  Maybe you need to push something to this branch.

tacaswell commented:
Sorry, meant appveyor, just having issues keeping them all straight :sheep:

JanSchulz commented:
@mdboom Thanks! Currently I have a appveyor build my master branch. Not sure what it does with PRs... We'll see on my next push :-)

Latest problem:

on 2.7, 32bit:
```
running build_ext
building 'matplotlib.ft2font' extension
creating build\temp.win32-2.7
creating build\temp.win32-2.7\Release
creating build\temp.win32-2.7\Release\src
C:\Users\appveyor\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -DFREETYPE_BUILD_TYPE=system -DPY_ARRAY_UNIQUE_SYMBOL=MPL_matplotlib_ft2font_ARRAY_API -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -IC:\Python27_32\envs\_build\lib\site-packages\numpy\core\include -I. -IC:\Python27_32\envs\_build\include -IC:\Python27_32\envs\_build\PC /Tpsrc/ft2font.cpp /Fobuild\temp.win32-2.7\Release\src/ft2font.obj
ft2font.cpp
C:\Users\appveyor\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Include\xlocale(342) : warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc 
c:\python27_32\conda-bld\work\src\ft2font.h(10) : fatal error C1083: Cannot open include file: 'ft2build.h': No such file or directory 
error: command 'C:\\Users\\appveyor\\AppData\\Local\\Programs\\Common\\Microsoft\\Visual C++ for Python\\9.0\\VC\\Bin\\cl.exe' failed with exit status 2
 
[test-environment] C:\Python27_32\conda-bld\work>if errorlevel 1 exit 1 
C:\Python27_32\conda-bld\work
```

on 3.5, 64bit: 
```
running build_ext
building 'matplotlib.ft2font' extension
creating build\temp.win-amd64-3.5
creating build\temp.win-amd64-3.5\Release
creating build\temp.win-amd64-3.5\Release\src
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MT -DFREETYPE_BUILD_TYPE=system -DPY_ARRAY_UNIQUE_SYMBOL=MPL_matplotlib_ft2font_ARRAY_API -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -IC:\Python35_64\envs\_build\lib\site-packages\numpy\core\include -I. -IC:\Python35_64\envs\_build\include -IC:\Python35_64\envs\_build\include -IC:\Python35_64\envs\_build\Library\include /EHsc /Tpsrc/ft2font.cpp /Fobuild\temp.win-amd64-3.5\Release\src/ft2font.obj
ft2font.cpp
src/ft2font.cpp(5): fatal error C1083: Cannot open include file: 'string': No such file or directory 
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\cl.exe' failed with exit status 2
 
[test-environment] C:\Python35_64\conda-bld\work>if errorlevel 1 exit 1 
```

tacaswell commented:
Does anything in https://github.com/jbmohler/matplotlib-winbuild help?

JanSchulz commented:
@tacaswell The normal wheel build works, I think it's something which is missing in the conda build environment...

JanSchulz commented:
@mdboom: can you post a link to the project page on appveyor? Mine looks like https://ci.appveyor.com/project/JanSchulz/matplotlib

tacaswell commented:
I seem to have attached it to my account by randomly pressing buttons... https://ci.appveyor.com/project/tacaswell/matplotlib

mdboom commented:
> I seem to have attached it to my account by randomly pressing buttons... https://ci.appveyor.com/project/tacaswell/matplotlib

I also did the same, I guess.  It's set up for the `matplotlib/matplotlib` repo, but it puts it under `project/mdboom/matplotlib`.

mdboom commented:
Looks like appveyor is reporting back here now.  Progress!

jenshnielsen commented:
Unfortunatly it's also reporting back on other prs i.e. #5591

JanSchulz commented:
Fun: https://github.com/matplotlib/matplotlib/commit/349762b8d2fba9638746e6ed1c8d237a925561ec introduced a depency on functools32 for py27, but conda has not yet such a package although such a package exists in their conda-recipe repo... Ok, so no 2.7 packages...

mdboom commented:
I'll report the functools32 issue.

JanSchulz commented:
Ok, I have no clue what this is:


py35, 64bit:
```
ft2font_wrapper.obj : warning LNK4197: export 'PyInit_ft2font' specified multiple times; using first specification
   Creating library build\temp.win-amd64-3.5\Release\src\ft2font.cp35-win_amd64.lib and object build\temp.win-amd64-3.5\Release\src\ft2font.cp35-win_amd64.exp
LINK : warning LNK4098: defaultlib 'MSVCRT' conflicts with use of other libs; use /NODEFAULTLIB:library
freetype.lib(ftrfork.c.obj) : error LNK2001: unresolved external symbol __imp_strrchr
freetype.lib(truetype.c.obj) : error LNK2001: unresolved external symbol __imp_strstr
freetype.lib(type1.c.obj) : error LNK2001: unresolved external symbol __imp_memchr
freetype.lib(sfnt.c.obj) : error LNK2001: unresolved external symbol __imp_memchr
build\lib.win-amd64-3.5\matplotlib\ft2font.cp35-win_amd64.pyd : fatal error LNK1120: 3 unresolved externals
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\link.exe' failed with exit status 1120
 ```


This seems to be reported in python upstream: https://bugs.python.org/msg248651 https://bugs.python.org/issue24872

mdboom commented:
Just in case it hasn't been said explicitly: thanks so much for taking this on, @JanSchulz.  At a recent meetup at PyData NYC, it was made pretty clear that the percentage of Windows use in our developer pool is approaching zero while the percentage of Windows among our users is easily more than half and growing as scientific Python makes its way into corporate environments.  So anything we can do to keep things working better on Windows is greatly appreciated.

WeatherGod commented:
Btw, when this gets all cleared up and kinks ironed out, we should probably
do the same for Basemap as I have *zero* Windows build expertise and I
wouldn't know the first thing on how to build a windows wheel.

On Wed, Dec 2, 2015 at 2:28 PM, Michael Droettboom <notifications@github.com
> wrote:

> Just in case it hasn't been said explicitly: thanks so much for taking
> this on, @JanSchulz <https://github.com/JanSchulz>. At a recent meetup at
> PyData NYC, it was made pretty clear that the percentage of Windows use in
> our developer pool is approaching zero while the percentage of Windows
> among our users is easily more than half and growing as scientific Python
> makes its way into corporate environments. So anything we can do to keep
> things working better on Windows is greatly appreciated.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/5604#issuecomment-161406736>
> .
>

JanSchulz commented:
@mdboom Ok, the functool32 addition in conda-recipe is a recent addition: https://github.com/conda/conda-recipes/pull/459

I really wonder how up-to-date conda-recipe is and how much magic is installed in anacondas autobuilder to let the matplotlib recipe produce the py35 packages of matplotlib :-/

cgohlke commented:
FWIW, https://bugs.python.org/msg248651 and https://bugs.python.org/issue24872 are obsolete and do not apply to Python 3.5 final.

JanSchulz commented:
@cgohlke thanks for the clarification!

back to the drawing board for py35... :-/ But only after 2.7 and 3.3 are running :-)

jenshnielsen commented:
As far as I know Python 3.5 is build with visual studio 2015 and not visual studio 2010 so perhaps that is the reason?

JanSchulz commented:
One strange thing in the setup process: the checks use `get_include_dirs()`, which uses `[d+"/include" for d in get_base_dirs()` and extends that with `os.environ.get('CPLUS_INCLUDE_PATH', '').split(os.pathsep))`. So if a header file is found in that dirs, the check passes.

On the other hand, the `PkgConfig.setup_extension()`, which does the actuall setup for the compiler, does **not** use  `os.environ.get('CPLUS_INCLUDE_PATH', '').split(os.pathsep))`, so fails during compilation...

JanSchulz commented:
Ok, this should build on three pys: py27,32|64bit and py34,64bit. It runs the tests and builds wheels on all platforms. On py3.4, it also builds a conda package.

If this gets green I would appreciate a rough review before I start polishing it (see todo...)...

JanSchulz commented on setupext.py:
```diff
@@ -318,6 +324,8 @@ def setup_extension(self, ext, package, default_include_dirs=[],
                     if os.path.exists(dir):
                         ext.include_dirs.append(dir)
+                ext.include_dirs.extend(
+                        os.environ.get('CPLUS_INCLUDE_PATH', '').split(os.pathsep))
```
This results in getting the include paths included twice... Need to investigate...

mdboom replied:
Is CPLUS_INCLUDE_PATH getting set by something?  I thought that was a gcc/Unixism...

JanSchulz replied:
I used it as a way to influence the base dirs without editing (=patching during build) setup.cfg and wondered why it didn't work...

mdboom replied:
Ok.  In that case, I think this block right here needs to be Windows-only.  It is redundant on Unix.

mdboom commented:
> On the other hand, the PkgConfig.setup_extension(), which does the actuall setup for the compiler, does not use os.environ.get('CPLUS_INCLUDE_PATH', '').split(os.pathsep)), so fails during compilation...

Sorry -- that was done with gcc in mind.  `gcc` always follows `CPLUS_INCLUDE_PATH`, so it doesn't need those directories explicitly passed to it, which is why matplotlib doesn't.  I guess VS doesn't use those at all.  I guess the `CPLUS_INCLUDE_PATH` stuff should be inside a `if sys.platform != 'win32'` block.

tacaswell commented:
It is _working_, but not reporting failures if the tests do not pass.

```python
FAILED (KNOWNFAIL=1213, SKIP=10, errors=3)
%CMD_IN_ENV% python setup.py bdist_wheel
Configuring Windows SDK "v7.1" for Python "3" on a 64 bit architecture

```

QuLogic commented:
Failures are [explicitly ignored](https://github.com/matplotlib/matplotlib/pull/5604/files#diff-180360612c6b8c4ed830919bbb4dd459R78), I think

tacaswell commented:
:+1: 

Might make sense to move the changes to setup.py for conda into a conda patch.

JanSchulz commented:
@tacaswell I wanted to have test packages. IMO the tests should once be in a "known good" state until that is enabled. I will file issues about it.

Will split the conda specific setup.py changes as a patch during conda build.

JanSchulz commented:
For the basedir thingy (`CPLUS_INCLUDE_PATH`): will removemy change, add a check for gcc to the current hack and then add a way to influence the basedirs via env variables. That should make this easier to build without setting up a setup.cfg...

JanSchulz commented:
I tried to include the TkAgg backend, but I don't understand why matplotlib setupext and conda use different tk versions:

matplotlib says, for 3.3+, you need tk8.6
https://github.com/matplotlib/matplotlib/commit/edb8cb35c281239a61c27dcfd3acfeb0a55a9ace

But conda comes with tk8.5:

```
matplotlib 1.4.3 np110py35_3 # same for 2.7 and so on...
[...]
dependencies:
[...]
    tk 8.5.*
```

Is that a conda speciality or does it not matter anymore?

mdboom commented:
I think conda just ships with a different Tk than stock Python.  conda heavily patches matplotlib to get their Tk -- that's in the "public" conda-recipe, and we can do something similar here, or ideally merge it upstream (without breaking current non-conda behavior).

JanSchulz commented:
AFAIU, the conda matplotlib recipe which is available is not the one which builds the matplotlib packages... E.g. the matplotlib package depends on tk and zlib, but the available recipe does not declare these dependencies...

mdboom commented:
> AFAIU, the conda matplotlib recipe which is available is not the one which builds the matplotlib packages... E.g. the matplotlib package depends on tk and zlib, but the available recipe does not declare these dependencies...

No -- they aren't the same.  But the conda-recipe has the patch to find libtk even though Tkinter isn't a dependency.

JanSchulz commented:
the last commits should fix the test failures on windows (e.g https://github.com/matplotlib/matplotlib/pull/5460 and the problem with NamedTempfile see here: https://github.com/matplotlib/matplotlib/issues/5610)

JanSchulz commented:
@tacaswell @mdboom One of you should probably remove the appveyoir integration: currently both of you run builds for each PR:
https://ci.appveyor.com/project/mdboom/matplotlib/history
https://ci.appveyor.com/project/tacaswell/matplotlib/history

github shows me the mdboom variant, but at some point I got the link to tacaswell.

tacaswell commented:
@mdboom I deleted mine

JanSchulz commented:
Yay, appveyor and travis are green...

tacaswell commented:
@JanSchulz I am inclined to merge this now that the tests are passing so the _rest_ of the PRs can report success on windows and iterate on the build products in a new PR.

JanSchulz commented:
I'm all for it... I would add the rest of the todos to another issue/PR and get them fixed there... I've also build functools32 and matplotlib in my package_builder repo and the packages are currently uplaoding: https://anaconda.org/janschulz I will update the conda packaging for osx/linux from that, so one could build conda packags via `conda build ci/conda_recipe` (and doing some steps to set the env vars beforehand...)

tacaswell commented:
:+1: thanks!

Should probably back-port this to 2.0.x ? @mdboom

JanSchulz commented:
If someone wants to test it: https://anaconda.org/janschulz/matplotlib/files

conda install -c https://conda.anaconda.org/janschulz matplotlib

[These are packages form my package builder repo, so they are available in a conda repo (for osx+win+linux and not win-only in the appveyor artefact). On the other hand: no wheels...]

pelson commented on ci/conda_recipe/condaversion.patch:
```diff
@@ -0,0 +1,14 @@
+ __version__ = versioneer.get_version()
+ 
++# For conda builds...
++with open("__conda_version__.txt", "w") as f:
```
This is not necessary (and needs to be removed). Runtime determination of the version can be done with:

```
{% set data = load_setuptools() %}
package:
   name: my_package
   version: {{ data.get('version') }}
```

JanSchulz replied:
Ui, first time I see that way to specify the version. I used the docs to get to that patch: http://conda.pydata.org/docs/building/meta-yaml.html#package-version

Why the "needs to be removed"?

pelson replied:
> Why the "needs to be removed"?

I hope it didn't come across as rude - I just wanted to be clear that if we remove this form, we fix the problem you describe in the recipe whereby:
```
# Note also that our setup.py script, which is called by conda-build, writes
+  # a __conda_version__.txt file, so the version number on the binary package
+  # is set dynamically. This unfortunately mean that conda build --output
+  # doesn't really work.
```.

pelson commented on ci/conda_recipe/meta.yaml:
```diff
@@ -0,0 +1,61 @@
+
+package:
+  name: matplotlib-test
+  version: 1.5.0.9
```
See my previous comment about runtime determination of this.

pelson commented on ci/conda_recipe/meta.yaml:
```diff
@@ -0,0 +1,61 @@
+    # don't run nosetest
+    - cfg_notests.patch
+    # nake Qt4Agg the default backend
+    - cfg_qt4agg.patch    [not osx]
```
These should be commented selectors.

pelson commented on lib/matplotlib/animation.py:
```diff
@@ -551,6 +553,27 @@ def delay(self):
+                binpath += '\\convert.exe'
+                break
+            except Exception:
+                binpath = ''
```
This doesn't sound very healthy.

pelson commented on lib/matplotlib/animation.py:
```diff
@@ -551,6 +553,27 @@ def delay(self):
+        rcParams[cls.exec_key] = rcParamsDefault[cls.exec_key] = binpath
+
+
+ImageMagickBase._init_from_registry()
```
This is a strange addition. I'm not sure why it is necessary.

jenshnielsen replied:
See #5460 for the discussion of this on windows

pelson commented on appveyor.yml:
```diff
@@ -0,0 +1,104 @@
+  - SET PATH=%PYTHON%;%PYTHON%\Scripts;%PATH%
+  - cmd: conda config --set show_channel_urls yes
+  # for msinttypes
+  - cmd: conda config --add channels conda-forge
```
Given you're already adding this, it would make some sense to rationalise the install and ``run_with_env`` code by using the ObviousCI version provided.

JanSchulz commented:
The curently unsolved windows build issues are now handled in https://github.com/matplotlib/matplotlib/pull/5629

@pelson I added you comments as Todos
